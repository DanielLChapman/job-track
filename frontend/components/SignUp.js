import gql from "graphql-tag";
import { useMutation } from "@apollo/client";
import useForm from "../library/useForm";
import { CURRENT_USER_QUERY } from "./User";
import { useState } from "react";

export const SIGNUP_MUTATION = gql`
    mutation SIGNUP_MUTATION(
        $name: String!
        $email: String!
        $password: String!
    ) {
        createUser(data: { name: $name, email: $email, password: $password }) {
            id
            name
            email
        }
    }
`;

function capitalize(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
}

function SignUp(props) {
    let [submitError, setError] = useState("");

    const { inputs, handleChange, resetForm } = useForm({
        email: "",
        password: "",
        name: "",
    });
    const [signup, { data, error, loading }] = useMutation(SIGNUP_MUTATION);
    const handleSubmit = async (e) => {
        e.preventDefault();
        // send email and password to graphql api
        let variables = {
            email: inputs.email,
            password: inputs.password,
            name: inputs.name,
        };

        if (inputs.password.length < 8) {
            setError("Password required to be a minimum of 8 characters");
            return;
        }

        let errorMessage;
        const res = await signup({
            variables: {
                email: inputs.email,
                password: inputs.password,
                name: inputs.name,
            },
        }).catch((error) => {
            //let a = error.graphQLErrors[0].message.splice(0,6);

            errorMessage = error.graphQLErrors[0];
            let a;
            let b;
            try {
                a = error.graphQLErrors[0].message.substring(0, 6);
                b = Object.keys(errorMessage.extensions.exception.keyValue)[0];
            } catch (e) {
                console.log(error.graphQLErrors[0]);
                if (error.graphQLErrors[0].extensions.prisma.code === "P2002") {
                    return setError(
                        "Email already used. Please login or try a different email"
                    );
                }
            }
            if (a === "E11000") {
                return setError(
                    `${capitalize(b)} is already taken, please choose a new one`
                );
            }
        });
        if (errorMessage) {
            console.log(errorMessage);
            return;
        }
        if (props.closeFunc) {
          props.closeFunc(props.closeValue);
        }
        resetForm();
    };

    return (
        <form method="POST" onSubmit={handleSubmit}>
            <h2>Sign Up For An Account</h2>
            {submitError ? <div>{submitError}</div> : <></>}
            <fieldset>
                {data?.createUser && (
                    <p data-testid="success-signup-message">
                        Signed Up With {data.createUser.email} - Please Go Ahead
                        And Sign In
                    </p>
                )}
                
                <div className="input">
                        <input
                            type="name"
                            name="name"
                            placeholder="Name"
                            autoComplete="name"
                            value={inputs.name}
                            onChange={handleChange}
                            className="input_field"
                            required
                        />
                        <label className="input_label" htmlFor="name">Name</label>
                    </div>
                <div className="input">
                        <input
                            type="email"
                            name="email"
                            placeholder="Your Email Address"
                            autoComplete="email"
                            value={inputs.email}
                            onChange={handleChange}
                            className="input_field"
                            required
                        />
                        <label className="input_label" htmlFor="email">Email</label>
                    </div>
                    <div className="input">
                        <input
                            type="password"
                            name="password"
                            placeholder="Password"
                            autoComplete="password"
                            value={inputs.password}
                            onChange={handleChange}
                            className="input_field"
                            required
                        />
                        <label className="input_label" htmlFor="password">Password</label>
                    </div>
                <button type="submit">Sign Up</button>
            </fieldset>
        </form>
    );
}

export default SignUp;
