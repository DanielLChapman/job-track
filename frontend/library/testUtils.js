import casual from 'casual';

casual.seed(231313);

const fakeUser = (overrides) => ({
    __typename: 'User',
    id: '4234',
    name: casual.name,
    email: casual.email,
    ...overrides,
  });

  export {
    fakeUser
  }