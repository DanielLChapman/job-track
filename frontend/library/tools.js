export const convertNum = (num) => {
    let t = num.toString();
    if (t.length < 3) {
        t = "$" + t
        return t;
    }
    let result = [];
    let count = 1;
    for (let i = t.length - 1; i >= 0; i--) {
        result.unshift(t[i]);
        
        if (count % 3 === 0 && i !== 0) {
            result.unshift(',')
        }
        count++;
    }

    result.unshift('$')

    return result.join('');

}