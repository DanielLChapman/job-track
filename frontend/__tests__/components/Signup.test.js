import {render, screen} from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';

import wait from 'waait';
import SignUp, {SIGNUP_MUTATION} from '../../components/SignUp';
import { fakeUser } from '../../library/testUtils';
import {act} from 'react-dom/test-utils';

const me = fakeUser();
const password = 'tempPassword';
const mocks = [
    {
        request: {
          query: SIGNUP_MUTATION,
          variables: {
            name: me.name,
            email: me.email,
            password,
          },
        },
        result: {
          data: {
            createUser: {
              __typename: 'User',
              id: 'abc123',
              email: me.email,
              name: me.name,
            },
          },
        },
      },
]


describe('<SignUp/>', () => {
    it('render and matches snapshot', () => {
      const { container } = render(
        <MockedProvider>
          <SignUp />
        </MockedProvider>
      );
      expect(container).toMatchSnapshot();
    });
    it('calls the mutation properly', async () => {
      const { container, debug } = render(
        <MockedProvider mocks={mocks}>
          <SignUp />
        </MockedProvider>
      );
      // Type into the boxes
      await userEvent.type(screen.getByPlaceholderText(/name/i), me.name);
      await userEvent.type(screen.getByPlaceholderText(/email/i), me.email);
      await userEvent.type(screen.getByPlaceholderText(/password/i), password);
      // Click the submit
      await userEvent.click(screen.getByText('Sign Up'));
      await act(async () => {
        await wait(0);
      })
      
      const element = await screen.getByTestId('success-signup-message');
      expect(element);
      /*await screen.findByText(
        `Signed Up With`
      );*/
    });
  });