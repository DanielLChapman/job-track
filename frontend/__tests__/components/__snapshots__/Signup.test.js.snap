// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp/> render and matches snapshot 1`] = `
<div>
  <form
    method="POST"
  >
    <h2>
      Sign Up For An Account
    </h2>
    <fieldset>
      <div
        class="input"
      >
        <input
          autocomplete="name"
          class="input_field"
          name="name"
          placeholder="Name"
          required=""
          type="name"
          value=""
        />
        <label
          class="input_label"
          for="name"
        >
          Name
        </label>
      </div>
      <div
        class="input"
      >
        <input
          autocomplete="email"
          class="input_field"
          name="email"
          placeholder="Your Email Address"
          required=""
          type="email"
          value=""
        />
        <label
          class="input_label"
          for="email"
        >
          Email
        </label>
      </div>
      <div
        class="input"
      >
        <input
          autocomplete="password"
          class="input_field"
          name="password"
          placeholder="Password"
          required=""
          type="password"
          value=""
        />
        <label
          class="input_label"
          for="password"
        >
          Password
        </label>
      </div>
      <button
        type="submit"
      >
        Sign Up
      </button>
    </fieldset>
  </form>
</div>
`;
